VAT Id Number validators for all EU countries

homebeaver opened the pull request:
Hi @garydgregory ,

here is EUgen from Europe. 
With this PR I provide **validation** and check digit **calculation** for VAT Id Numbers [VATIN](https://en.wikipedia.org/wiki/VAT_identification_number) for all countries in the EU. This number is neccessary in the intra comunity trades to apply a zero VAT rate. 
You can find a documentation of the different VATINs structures in my [VATIN wiki (de)](https://github.com/homebeaver/ungueltig/wiki/VATIN)

The validator for all is `VATINValidator` - it implements the [jira issue 494](https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494)

Pls review and merge.
Regards EUGen

homebeaver committed:
add SIRENE Validator: SIREN and SIRET

homebeaver committed:
Merge remote-tracking branch 'apache/master' into sirene

homebeaver committed:
Merge remote-tracking branch 'homebeaver/sirene' into master+vatin

homebeaver committed:
Merge commit 'ca4db8dca6829bce1b1701a74a115c6e2e39f2f0' into master+vatin

homebeaver committed:
Merge commit '5335159022d750558c4e5a46028cc8467586b05b' into master+vatin

homebeaver committed:
Merge commit '4a87127ac38047a70aaf522ae105701ff5343237' into master+vatin

homebeaver committed:
Merge commit '682acf26359eabcac1b942ef8ec807a3e4033fb8' into master+vatin

homebeaver committed:
Merge commit 'd3c9cf0457c1bbb6256d3b0e7cbbc8203d4de419' into master+vatin

homebeaver committed:
Merge commit '5b3bc4bdb674490419a62559cd222819b93750eb' into master+vatin

homebeaver committed:
Merge commit '77951ab6ba08753488e88385209663573fc00ea9' into master+vatin

homebeaver committed:
Merge commit 'c05298e773b1bb570d545d85fd10d8f203b25309' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
add Austrian VAT identification number (VATIN)

Check Digit calculation/validation

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
reported by Checkstyle

homebeaver committed:
Re-run Maven using the -X (--debug) switch to get info for ...

Error:  Failed to execute goal org.apache.maven.plugins:maven-pmd-plugin:3.24.0:check (default-cli) on project commons-validator: PMD 7.4.0 has found 1 violation. For more details see: /home/runner/work/commons-validator/commons-validator/target/pmd.xml -> [Help 1]

homebeaver committed:
Revert "Re-run Maven using the -X (--debug) switch to get info for ..."

This reverts commit 62c5f75251ce3952eacace924b1225d4156a9998.

homebeaver committed:
checkstyle

homebeaver committed:
comment

homebeaver committed:
add common message texts to CheckDigitException

minor changes in VATidATCheckDigit

homebeaver committed:
checkstyle

homebeaver committed:
add Belgian VAT identification number (VATIN) - first shot

Check Digit calculation/validation
It is a subclass from Modulus97CheckDigit, which is also used to
validate the LEI and German Leitweg, see
https://github.com/apache/commons-validator/pull/61

This implemets parts of
https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
minor change in VATidATCheckDigit

homebeaver committed:
VATidBECheckDigit extends Modulus97CheckDigit

homebeaver committed:
AbstractCheckDigitTest.createInvalidCodes for two char check digits

homebeaver committed:
checkstyle

homebeaver committed:
error occurred in Javadoc report

homebeaver committed:
add Bulgarian VAT identification number (VATIN)

Check Digit calculation/validation

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
refine code for Bulgarian VATIN - TODO remove java.logging

- add testcases with branches

homebeaver committed:
add testdata from from https://de.wikipedia.org/wiki/Leitweg-ID

homebeaver committed:
Bulgarian VATIN for physical persons (civil number)

homebeaver committed:
use apache.commons.logging

homebeaver committed:
add Cypriot VATIN check digit validator and test (first shot)

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
code housekeeping for Cypriot VAT identification number (VATIN)

homebeaver committed:
add Czech VATIN check digit validator and test (first shot)

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
common method for Czech: Rodné číslo (RČ)

homebeaver committed:
Housekeeping Czech VATIN

homebeaver committed:
use apache.commons.logging

homebeaver committed:
PMD check UselessParentheses

homebeaver committed:
add German VAT identification number (VATIN)

Check Digit calculation/validation is implemented in
Modulus11TenCheckDigit which is also used for German Tax Id Number (TIN)
and Croatia VATIN

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
Housekeeping DE VATIN Check Digit

TODO remove VATidDECheckDigit, because superclass Modulus11TenCheckDigit
implements all

homebeaver committed:
Remove Modulus11TenCheckDigit

homebeaver committed:
VATidDECheckDigitTest more Tests

homebeaver committed:
improve MOD 11,10 module (Modulus11TenCheckDigit),

add HR VAT Id Check Digit Tests, which uses Modulus11TenCheckDigit

add German Tax ID Check Digit and -test

homebeaver committed:
housekeeping

homebeaver committed:
add Danish VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
housekeeping

homebeaver committed:
add Estonian VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
housekeeping

homebeaver committed:
add Greek VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
housekeeping EL VATIN

homebeaver committed:
add Spanish VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
Spanish VATIN housekeeping, use apache.commons.logging

homebeaver committed:
add Finnish VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
add a commented test

homebeaver committed:
add French VATIN check digit validator and test (first shot)

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
check SIREN in French VAT identification number

There are more than one possible VATIN check digit C(1-2) for a given
SIREN, thus, it isn't possible to calculate the right one.

homebeaver committed:
PMD rule="UselessParentheses" ruleset="Code Style"

homebeaver committed:
add HU VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
add Irish VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
add Italian VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

TODO remove informational empty class VATidITCheckDigit because IT use
LuhnCheckDigit, add Tests to LuhnCheckDigitTest

homebeaver committed:
remove informational empty class VATidITCheckDigit because

IT use LuhnCheckDigit, add Tests to LuhnCheckDigitTest

homebeaver committed:
Irish VATIN housekeeping

homebeaver committed:
add Lithuanian VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
Lithuanian VATIN housekeeping

homebeaver committed:
add Luxemburg VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
add Latvian VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
calculate NMIN (TIN) check digit for natural persons

homebeaver committed:
Latvian VATIN housekeeping

homebeaver committed:
checkstyle

homebeaver committed:
add Malta VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
method Modulus97CheckDigit.toCheckDigit : change visibility

homebeaver committed:
Malta VATIN housekeeping

homebeaver committed:
add Netherlands VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
Netherlands VATIN housekeeping

homebeaver committed:
add Polish VATIN check digit validator and test (first shot)

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
SireneValidatorTest : add testcase from overseas department

homebeaver committed:
Polish VAT identification number (VATIN) housekeeping

homebeaver committed:
add Netherlands VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
Portugal VATIN housekeeping

homebeaver committed:
add Romanian VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
Romanian VATIN housekeeping

homebeaver committed:
add Swedish VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
housekeeping for Swedish VATIN

homebeaver committed:
add Slovenian VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
housekeeping Slovenian VAT

homebeaver committed:
add Slovakia VATIN check digit validator and test

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
add UK VATIN check digit validator and test (first shot)

UK is not part of EU, but Northern Ireland (XI)

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
VATINCheckDigit check digit validator class: a mapping for all ...

EU  countries - TODO housekeeping

https://issues.apache.org/jira/projects/VALIDATOR/issues/VALIDATOR-494

homebeaver committed:
VATINCheckDigit comment and housekeeping

homebeaver committed:
Swedish VATIN : CheckDigit LUHN_CD_INSTANCE

homebeaver committed:
VAT identification number (VATIN) Validator / main routine (first shot)

homebeaver committed:
TidDECheckDigit use apache.commons.logging

homebeaver committed:
test spaces in SIRET overseas department

homebeaver committed:
work in progress for FR-VATIN

use Modulus97CheckDigit.toCheckDigit

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
work in progress : VATIN Validator

homebeaver committed:
comments and tests

homebeaver committed:
remove the prefix "EU" Foreign (non EU) companies

homebeaver committed:
remove otherCountryCodes from VATINValidator.Validator

homebeaver committed:
AbstractCheckDigitTest.createInvalidCodes provide to create codes with

checkdigit length > 1 and checkdigit as prefix

The default is checkdigit as suffix : protected method createCode
returns code + cd

this can be overridden to cd + code

homebeaver committed:
simplify VATINValidator with formats and check routines for EU countries

VATINs for 28 EU countries can be validated

homebeaver committed:
forgot to checkin VATINValidatorTest

homebeaver committed:
VATidGBCheckDigit for Northern Ireland - work in progress

homebeaver committed:
VATidXICheckDigitTest new style 9755 tests

homebeaver committed:
style check

homebeaver committed:
Revert "style check"

This reverts commit 9159c4f0c343f193633e90f75f679aa6586c11cc.

homebeaver committed:
wrong check in revert + commit VATidXICheckDigitTest

homebeaver committed:
finish GB+XI VATINS validating also codes with branch ids

homebeaver committed:
finish VATidFRCheckDigit

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver commented:
> Is it really necessary to have a separate class for each country? Also DE has a test class but no main class

@sebbASF 

Yes. Each of the 28 countries use a different algoritm to calculate/validate the check digit. 
Germany and Croatia use a standard **ISO/IEC 7064, MOD 11,10**. So the tests are against Modulus11TenCheckDigit
Italy uses LUHN.

homebeaver committed:
IT VAT Id Check Digit Tests against LUHN

homebeaver committed:
Speling

tkx to https://github.com/sebbASF

homebeaver commented on src/main/java/org/apache/commons/validator/routines/checkdigit/CheckDigitException.java:
```diff
@@ -25,6 +25,16 @@ public class CheckDigitException extends Exception {
+     * Common message text when code sum is zero
+     */
+    public static final String ZREO_SUM = "Invalid code, sum is zero";
+
```
Resolved with commit 547c10ee9023600a3547883a1394c8e7abce303f

garydgregory commented on src/main/java/org/apache/commons/validator/routines/checkdigit/Modulus97CheckDigit.java:
```diff
@@ -0,0 +1,177 @@
+ *
+ * @since 1.10.0
+ */
+public class Modulus97CheckDigit extends AbstractCheckDigit implements Serializable {
```
Don't implement Serializable, we are moving away from that IMO, folks that need serialization of some kind can use serialization proxies; see "Effective Java" by J. Block.

homebeaver replied:
Cannot remove it.
Error:  Failures: 
Error:    Modulus97CheckDigitTest>AbstractCheckDigitTest.testSerialization:300 org.apache.commons.validator.routines.checkdigit.Modulus97CheckDigit error during serialization: java.io.NotSerializableException: org.apache.commons.validator.routines.checkdigit.Modulus97CheckDigit

garydgregory replied:
Hello @homebeaver 
That was a bug in our test ;-) Please rebase on master to pick up https://github.com/apache/commons-validator/commit/fdc8ee9c26f668192b34ec5b955b3adf47f98b2d

homebeaver replied:
after rebase and rempve implement Serializable:
```
[INFO] --- spotbugs-maven-plugin:4.8.6.4:check (default-cli) @ commons-validator ---
[INFO] BugInstance size is 1
[INFO] Error size is 0
[INFO] Total bugs: 1
Error:  Medium: Class org.apache.commons.validator.routines.CodeValidator defines non-transient non-serializable instance field checkdigit [org.apache.commons.validator.routines.CodeValidator] In CodeValidator.java SE_BAD_FIELD
[
```

garydgregory replied:
Ah, I see: Because `CodeValidator implements Serializable` and `CodeValidator` has a `CheckDigit` field, we are stuck with `CheckDigit` implementing `Serializable` if we want to honor the `Serializable` contract on `CodeValidator`. I guess we'll have to drop serialization in a major version bump if we want to go that way.

TY for cycling through this one @homebeaver ! 😄

homebeaver replied:
I revert the commit 8add96b5939ef1cc2d3ab3ff8d774ee48f194d73

garydgregory commented on src/main/java/org/apache/commons/validator/routines/checkdigit/VATidBECheckDigit.java:
```diff
@@ -0,0 +1,112 @@
+
+    /**
+     * {@inheritDoc}
+     * return expression overridden
```
Missing `@`

garydgregory commented on src/main/java/org/apache/commons/validator/routines/checkdigit/VATidBGCheckDigit.java:
```diff
@@ -0,0 +1,216 @@
+    }
+
+    /**
+     * there ate three length
```
Sentences start with a capital letter. If you expect the comment below to be renders line by line, use a `ol` or `ul` HTML tag.

garydgregory commented on src/main/java/org/apache/commons/validator/routines/checkdigit/VATidCYCheckDigit.java:
```diff
@@ -0,0 +1,135 @@
+     */
+    @Override
+    protected int weightedValue(final int charValue, final int leftPos, final int rightPos) {
+        if (leftPos % 2 == 0) return charValue;
```
Always use blocks.

garydgregory commented on src/main/java/org/apache/commons/validator/routines/checkdigit/VATidFRCheckDigit.java:
```diff
@@ -0,0 +1,198 @@
+    private static final SireneValidator SIRENE_VALIDATOR = SireneValidator.getInstance();
+
+    /**
+     * Constructs a Check Digit routine.
```
Not helpful IMO. Also, there is no "routine" concept in Java, this constructs a new instance if anything ;-)

garydgregory reviewed (changes requested):
Hello @homebeaver 
This is quite a giant PR! I will take time to review and consider. I've made some initial but low-level comments.

homebeaver committed:
SireneValidator minor changes tkx to https://github.com/garydgregory

homebeaver committed:
more small chenges in SireneValidator

homebeaver committed:
resolved files VATINValidator, Modulus11TenCheckDigit,

Modulus97CheckDigit, TidDECheckDigit, VATidBECheckDigit, BG, CY, FR

tkx to Gary

homebeaver committed:
implement Serializable, restore previous version

garydgregory commented on src/main/java/org/apache/commons/validator/routines/checkdigit/TidDECheckDigit.java:
```diff
@@ -0,0 +1,119 @@
+    }
+
+    /**
+     * {@inheritDoc} <br>
```
Don't leave HTML tags open. You have paragraphs below, this break is likely not needed.

garydgregory commented on src/main/java/org/apache/commons/validator/routines/checkdigit/VATidLVCheckDigit.java:
```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    private int vRule1(String code) {
```
Use `final` where you can, throughout the PR.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/SireneValidatorTest.java:
```diff
@@ -0,0 +1,62 @@
+
+public class SireneValidatorTest {
+
+    private final String[] validFormat = new String[] {
```
If you define these test fixtures with `Arrays.asList(...)`, you can use `forEach()` to reduce clutter in the tests. YMMV.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/SireneValidatorTest.java:
```diff
@@ -0,0 +1,62 @@
+ */
+package org.apache.commons.validator.routines;
+
+import static org.junit.Assert.assertFalse;
```
Please use JUnit 5 `Assertions`, not JUnit 4 `Assert`.

Please rebase on git master where I finished porting to JUnit 5; there was only one class left to do.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/VATINValidatorTest.java:
```diff
@@ -0,0 +1,251 @@
+    @Test
+    public void testSetDefaultValidator1() {
+        final IllegalStateException thrown = assertThrows(IllegalStateException.class, () -> VALIDATOR.setValidator("GB", 15, "GB", null));
+        assertThat(thrown.getMessage(), is(equalTo("The singleton validator cannot be modified")));
```
Please, not Hamcrest, not AssertJ, Java is verbose enough as it is without cluttering tests which should be as simple as possible to understand since they also serve to some extent as examples. This is so much simpler to read and maintain:
```
 assertEquals("The singleton validator cannot be modified", thrown.getMessage());
 ```
There is absolutely no need for 2 extra method calls to assert this outcome.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/VATINValidatorTest.java:
```diff
@@ -0,0 +1,251 @@
+
+    private static final VATINValidator VALIDATOR = VATINValidator.getInstance();
+
+    // Eclipse 3.6 allows you to turn off formatting by placing a special comment, like
```
You don't need a comment about Eclipse. We use this `formatter` comment all over Commons 😉

garydgregory commented on src/test/java/org/apache/commons/validator/routines/VATINValidatorTest.java:
```diff
@@ -0,0 +1,251 @@
+    public void testSetValidatorLen35() {
+        final VATINValidator validator = new VATINValidator();
+        final IllegalArgumentException thrown = assertThrows(IllegalArgumentException.class, () -> validator.setValidator("DE", 35, "DE", null));
+//        System.out.println("testSetValidatorLen35 : thrown.getMessage():" + thrown.getMessage());
```
Remove comment.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/checkdigit/VATidFRCheckDigitTest.java:
```diff
@@ -0,0 +1,132 @@
+            , "40303265045"
+            , "23334175221"
+            , "06399859412"
+//            , "K7399859412" // SLRL ALTEA EXPERTISE COMPTABLE => test in VATINValidatorTest
```
Is this work in progress?

garydgregory commented on src/test/java/org/apache/commons/validator/routines/checkdigit/VATidFRCheckDigitTest.java:
```diff
@@ -0,0 +1,132 @@
+            , "06399859412"
+//            , "K7399859412" // SLRL ALTEA EXPERTISE COMPTABLE => test in VATINValidatorTest
+            , "83404833048"
+            , "11123456782" // konstruiert: L'identifiant n'existe pas dans le répertoire Sirene.
```
This comment needs to be more helpful to maintainers, as it the one above.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/checkdigit/VATidFRCheckDigitTest.java:
```diff
@@ -0,0 +1,132 @@
+            }
+            boolean res = routine.isValid(invalidCheckDigits[i]);
+            if (res) {
+//                log.warn("   " + i + " Testing Invalid Check Digit, Code=[" + invalidCheckDigits[i] + "] is true, expected false.");
```
Remove // commented out code, these look like work-in-progress.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/checkdigit/VATidLVCheckDigitTest.java:
```diff
@@ -0,0 +1,55 @@
+/**
+ * LV VAT Id Check Digit Tests.
+ */
+/*
```
Merge comments.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/checkdigit/VATidROCheckDigitTest.java:
```diff
@@ -0,0 +1,53 @@
+/**
+ * RO VAT Id Check Digit Tests.
+ */
+/*
```
Merge comments.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/checkdigit/VATidSKCheckDigitTest.java:
```diff
@@ -0,0 +1,54 @@
+/**
+ * SK VAT Id Check Digit Tests.
+ */
+/*
```
Merge comments.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/checkdigit/VATidXICheckDigitTest.java:
```diff
@@ -0,0 +1,173 @@
+/**
+ * XI VAT Id Check Digit Tests.
+ */
+/*
```
Merge comments.

garydgregory commented on src/test/java/org/apache/commons/validator/routines/checkdigit/VATidXICheckDigitTest.java:
```diff
@@ -0,0 +1,173 @@
+              , "888851256"
+              , "816137833"
+              , "562235945"
+              /* new style 9755 cannot be tested here
```
What does "new style" mean? Think about maintenance in these comments ;-)
Where can it be tested then?

homebeaver replied:
I`ve explained "new style" in the file. Here in short:
- each VAT code in GB and Nothern Ireland (XI) has two check digits - both are valid
- one calculated with MOD97 algorithm - this is called "old style" (this is the result of calculate method)
- the second calculated  with MOD9755 - this is called "new style"

garydgregory reviewed (changes requested):
Hello @homebeaver 

Thank you for you updates.

In general, where does the test data come from? If there are resources on the web, it would be nice to provide URLs in the Javadoc for provenance.

I've added comments scattered throughout.

Please rebase on git master.

TY!

garydgregory commented:
Hi @homebeaver 

A nice to have item would be an update to the `package-info.java` file for the package in play.

homebeaver committed:
Can use ABANumberCheckDigit for Estonian VATIN

homebeaver commented:
Hi @garydgregory 
where does the test data come from?

Different - I documented it in [wiki/VATIN](https://github.com/homebeaver/ungueltig/wiki/VATIN) - in German!
You'll find examples in the Details-link:

![grafik](https://github.com/user-attachments/assets/3ae433e6-c391-4f6e-8bf5-303fbaf78a11)

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Do not implement Serializable

homebeaver committed:
Revert "Do not implement Serializable"

This reverts commit 8add96b5939ef1cc2d3ab3ff8d774ee48f194d73.

homebeaver committed:
close HTML tag, final in method param

homebeaver committed:
use Lists validFormat, invalidFormat and forEach() in test

homebeaver committed:
checkstyle and error has occurred in Javadoc report generatio

homebeaver committed:
HTML in comment

homebeaver committed:
HTML blocks

homebeaver committed:
remove hamcrest, use List and forEach(...)

homebeaver committed:
comment special FR test cases

homebeaver committed:
Merge comments in tests LV RO SK

homebeaver committed:
XI VAT Id Check Digit Tests. Merge comments. Explain "new style"

Move new style 9755 test codes to VATINValidatorTest

homebeaver committed:
update to the package-info.java in org.apache.commons.validator.routines

homebeaver committed:
HTML tag

homebeaver commented:
> A nice to have item would be an update to the `package-info.java` file for the package

Hi @garydgregory 

updated the `package-info.java`  in `org.apache.commons.validator.routines
`
Regards - next week I'm on holiday :-)
EUGen

homebeaver committed:
Test, Comment and Simplify Estonian + Hungarien VATIN

homebeaver committed:
Merge comments in tests

homebeaver committed:
HTML Tags in comment

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Greek VATIN Testcase wich results to checkdigit 10 == 'X'

homebeaver committed:
test Slovenian VATIN

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
NL VATIN Testcase which results to checkdigit 10 == 'X'

homebeaver committed:
Method AbstractCheckDigitTest.createInvalidCodes should return empty

array if checkDigitLth is 0. This applays to VATIN in DK and SK.

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Revert "Method AbstractCheckDigitTest.createInvalidCodes should return empty"

This reverts commit 18741c648b88ba60c416a516b53db067fb23b9e5.

homebeaver committed:
Uups - commit reverted + merge master branch + correct commit ...

Method AbstractCheckDigitTest.createInvalidCodes should return empty
array if checkDigitLth is 0. This applays to VATIN in DK and SK.

homebeaver committed:
Modulus 11-X module simplifies a group of VATIN calculations

homebeaver committed:
Simplify VATIN for BG , LT , LV

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Luxemburg VATIN - accept "00" check digit

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
resolve errors reported by Checkstyle 10.18.2

homebeaver committed:
resolve remaining errors reported by checkstyle

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

raboof commented:
I'm afraid the NL validation is incomplete: my VAT ID, NL004350351B91, is not accepted while it is in fact valid (confirmed with the online VIES check). I suspect this may have to do with a change in 2020 or so: before that time, the btw-id for self-employed people consisted of 'NL'+their BSN (the Dutch SSN) + "B" + two digits. While 'leaking' a SSN isn't as problematic in NL as it is in some other countries, it's not great, so they stopped doing that. Apparently the BSN adheres to the `MOD-11#i` pattern, while the new number does not. I can share a second number that fails the validator but is correct according to VIES privately on request.

raboof commented:
> Is the new NL VAT validation scheme officially documented anywhere?

I looked around a bit but didn't find anything - the sources I found just say 'NL, 9 digits, B, 2 digits', no further structure described.

> It seems to me that even with access to official documentation, it is going to be a mammoth job keeping track of changes.
>
> Tracking updates to IBAN and Domain entries is time-consuming enough, and they each have a single authoritative source.
> 
> Sorry, but I think the VAT validation proposal is not a suitable candidate for Commons Validator.

OTOH, you could also say that this is a good motivation to share this effort. Perhaps we should make sure that there are at least 3 or so parties that are planning to use this in production (and share notes when rules change) before including it?

garydgregory commented:
I support @raboof's idea 👍

homebeaver commented:
> ... VAT ID, NL004350351B91, is not accepted while it is in fact valid (confirmed with the online VIES check). I suspect this may have to do with a change in 2020 or so: before that time, the btw-id for self-employed people consisted of 'NL'+their BSN (the Dutch SSN) + "B" + two digits. ...

@raboof thank you for testing. Indeed, my algorithm use MOD 11 with $${ G_i = i }$$ - this was changed in 2020: 
There are two groups with VAT-numbers:
• Modulo 11 VAT-ID numbers;
• Modulo 97 VAT-ID numbers.

I´ll correct this.

regards EUGen H.

homebeaver committed:
NL VATIN with MOD 97 used additionally since 2020

tkx to https://github.com/raboof for the hint

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver commented:
> I looked around a bit but didn't find anything ...

@raboof  - I found this [BMF_UID_Konstruktionsregeln-Nov-2020.pdf bmf.gv.at](https://www.bmf.gv.at/dam/bmfgvat/finanzonline/fuer-softwarehersteller/sonstige-erklaerungen/Zusammenfassende-Meldung/BMF_UID_Konstruktionsregeln_Stand_November-2020.pdf)

homebeaver commented:
> ... it does not check that the ID is actually in use. So I wonder what the use case is?

Compate VATID to IBAN: validating an IBAN does not check if the IBAN is actually in use.
Use VIES for checking if a VATIN is actually in use. It delegates the check to the national systems, but is not avaiable 7/24. We use it for valid VATINS only when creating an invioce.

The use cases: 
european standard EN16931-1:2017 for electronic invoices needs an valid VATIN of the cusomer. This number is neccessary in the intra comunity trades to apply a zero VAT rate.

In my case I migrate ERP data from system A to system B. There are a lot of unchecked VATINs and VIES is not the tool of choice for checking a mass. When a shop or company discontinues VIES returns invalid (not invalid since...), but invoices are stored a while for auditing.

homebeaver committed:
VATidNLCheckDigit : Line has trailing spaces + simplify

homebeaver committed:
checkstyle : whitespace

raboof commented:
> > I looked around a bit but didn't find anything ...
> 
> @raboof - I found this [BMF_UID_Konstruktionsregeln-Nov-2020.pdf bmf.gv.at](https://www.bmf.gv.at/dam/bmfgvat/finanzonline/fuer-softwarehersteller/sonstige-erklaerungen/Zusammenfassende-Meldung/BMF_UID_Konstruktionsregeln_Stand_November-2020.pdf)

Nice find! I checked and can confirm that both my own btw-id and the one I won't disclose are accepted by the updated implementation.

To justify the additional maintenance burden it'd still be good to find one or two other parties that can confirm they'd use this component in production and help keep it up-to-date before merging it into the main library.

sebbASF reviewed (changes requested):
These look like bugs

homebeaver commented on src/main/java/org/apache/commons/validator/routines/checkdigit/VATidBECheckDigit.java:
```diff
@@ -0,0 +1,104 @@
+    public String calculate(final String code) throws CheckDigitException {
+        if (GenericValidator.isBlankOrNull(code)) {
+            throw new CheckDigitException(CheckDigitException.MISSING_CODE);
+        }
```
You can find the explanation in `package-info` of  package `org.apache.commons.validator.routines.checkdigit`:


_This package contains Check Digit validation/calculation routines._

_Note that these **do not validate the input for length** or syntax. Such validation is performed by the org.apache.commons.validator.routines.XYZValidator classes._

`VATINValidator` class validates against reg-ex `"[0-1]\\d{9}"`

In checkdigit package I use LEN only if it is absolute necessary for the implementation of the check.

homebeaver replied:
To be consistent with `VATidBECheckDigit`

I change in `VATidLUCheckDigit` from `MIN_LEN` which is 4 to the stronger `CHECKDIGIT_LEN` which is 2.

Done in Commit dc6e9c1

homebeaver committed:
VATidLUCheckDigit class comment - tkx to https://github.com/sebbASF

homebeaver commented on src/main/java/org/apache/commons/validator/routines/checkdigit/VATidLUCheckDigit.java:
```diff
@@ -0,0 +1,110 @@
+ * Numéro d'identification à la taxe sur la valeur ajoutée {@code 123456pp}.
+ * </p>
+ * <p>
+ * The check digits are calculated as 89 - MOD 89
```
Resolved with Commit 5e1c022

homebeaver committed:
VATidLUCheckDigit : to be consistent with `VATidBECheckDigit`

I change in `VATidLUCheckDigit` from `MIN_LEN` which is 4 to the
stronger `CHECKDIGIT_LEN` which is 2

tkx to https://github.com/sebbASF

homebeaver commented:
> These look like bugs

No. Imprecise comments + inconsistent code. Both resolved. Tkx

homebeaver commented on src/main/java/org/apache/commons/validator/routines/checkdigit/VATidBECheckDigit.java:
```diff
@@ -0,0 +1,104 @@
+        }
+        final String check = code.substring(code.length() - CHECKDIGIT_LEN);
+        try {
+            final String cd = calculate(code.substring(0, code.length() - CHECKDIGIT_LEN)); // throws CheckDigitException
```
I plan a new PR with [implementation for all modules defined in ISO/IEC 7064](https://issues.apache.org/jira/browse/VALIDATOR-481) when this PR is finished/merged. 
This would change some of the base classes, anyway.

homebeaver commented on src/main/java/org/apache/commons/validator/routines/checkdigit/VATidBECheckDigit.java:
```diff
@@ -0,0 +1,104 @@
+    @Override
+    public String calculate(final String code) throws CheckDigitException {
+        if (GenericValidator.isBlankOrNull(code)) {
+            throw new CheckDigitException(CheckDigitException.MISSING_CODE);
```
I plan a new PR with [implementation for all modules defined in ISO/IEC 7064](https://issues.apache.org/jira/browse/VALIDATOR-481) when this PR is finished/merged. 
This would change some of the base classes, anyway.

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
avoid Javadoc Warnings: use of default constructor,

which does not provide a comment (implicit constructor)

garydgregory commented:
I wouldn't except any logging from this component.

homebeaver committed:
remove logging - not really needed.

https://github.com/apache/commons-validator/pull/271#issuecomment-2629127531

There are some classes with logging, but not all. ...
Is it really needed?

VATidCZCheckDigit : LOG was used only when DebugEnabled
VATidBGCheckDigit : dto
VATidESCheckDigit : dto
VATidFRCheckDigit : dto
VATidGBCheckDigit : dto
VATidLVCheckDigit : dto + to explain why calculation fails

homebeaver committed:
resolve pdm Code Style problem

homebeaver commented:
I've removed logging.

There is one class with logging left : `TidDECheckDigit`. The reason is the specification - the link to it is in the class header. The spec is in german. In some cases valid tids should not be used - and I warn with the message recomended in the spec. See page 6+7 of the spec or screenshot:
![grafik](https://github.com/user-attachments/assets/3eb43469-927b-4c8b-b793-ab250cea99df)

> I wouldn't except any logging from this component.

@garydgregory 
To satisfy you I can remove the class from this PR

sebbASF reviewed (changes requested):
As per comments

homebeaver committed:
removed class TidDECheckDigit

https://github.com/apache/commons-validator/pull/271#issuecomment-2629581560

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver committed:
cvc-elt.1.a error in pom.xml

- cvc-elt.1.a: Cannot find the declaration of element 'project'.
- Downloading external resources is disabled.

Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver commented:
Hi @sebbASF , @sebbASF 

when merging apache/master to my repo branch I got an error in `pom.xml`

```
- cvc-elt.1.a: Cannot find the declaration of element 'project'.
- Downloading external resources is disabled.
```

The solution is **https** instead http in pom element 'project'

I resolve this in my pull request and hope you accept this pull. It is waiting to be accepted for more then one year!

regards EUGen

homebeaver committed:
Revert "cvc-elt.1.a error in pom.xml"

This reverts commit 3b500c4bde3953d15b656d052d25c38aebebeb7b.

homebeaver committed:
Merge remote-tracking branch 'apache/master' into master+vatin

homebeaver commented:
tkx @raboof.
Changes can be cleanly merged. @garydgregory

garydgregory commented:
> > I resolve this in my pull request and hope you accept this pull. It is waiting to be accepted for more then one year!
> 
> As mentioned earlier in [#271 (comment)](https://github.com/apache/commons-validator/pull/271#issuecomment-2623175681): keeping this up-to-date looks like quite a maintenance burden. This means the work is valuable to share - but only if we're confident that there will be someone looking after this code after it's merged. I assume you're using this and would be ready to take some of that maintainership on yourself, but it'd be good to have one or two other people showing up.

There has been no reply from @homebeaver to your comment @raboof, which doesn't inspire confidence that this will be maintained.

garydgregory reviewed (commented):
@homebeaver At least one of my comments hasn't been addressed; it was marked resolved, which is not appropriate IMO, so I stopped looking. Beyond that, I'm not sure there is consensus to bring this in due to the maintenance burden. What do others think?

sebbASF commented:
I agree that this is likely to become too much of a maintenance burden.

Seems to me this would be better as a separate project (with its own package name).
This could of course use existing Validator methods where appropriate.

It might be worth considering adding some or all of the new ModulusCheckDigit classes, if the algorithm is standardised and in general use.
